server/util/cache/cache.go

```diff
@@ -0,0 +1,252 @@
+package cache
+
+import (
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/buildbuddy-io/buildbuddy/server/metrics"
+	"github.com/prometheus/client_golang/prometheus"
+
+	gstatus "google.golang.org/grpc/status"
+)
+
+type CacheLayer string
+
+const (
+	MemoryCacheLayer CacheLayer = "memory"
+	CloudCacheLayer  CacheLayer = "cloud"
+
+	setMethodLabel   = "single"
+	writeMethodLabel = "streaming"
+)
+
+// CacheMetrics exposes APIs for instrumenting cache operations.
+type CacheMetrics struct {
+	Layer   CacheLayer
+	Backend string
+}
+
+func (c *CacheMetrics) NewInstrumentedReader(r io.Reader) *InstrumentedReader {
+	return &InstrumentedReader{
+		Reader: r,
+		cacheMetrics: c,
+	}
+}
+
+func (c *CacheMetrics) NewInstrumentedWriteCloser(w io.WriteCloser) *InstrumentedWriteCloser {
+	return &InstrumentedWriteCloser{
+		WriteCloser: w,
+		cacheMetrics: c,
+	}
+}
+
```

Comment:
on paper these look OK but they also really scare me since they wrap the core read / write interfaces!

I was trying to think of "lighter" approaches -- one idea I had was that since the digest includes the number of bytes that are being transferred, we could just use the same CacheTimer approach, but allow passing in the number of bytes?

Like:
timer := NewCacheTimer("GCS", digest.GetSizeBytes())
timer.Close() <--- this actually writes the data

Then you could embed this in the gcsDedupingWriteCloser struct and call Close inside of Close().

The reader case you do still need to make a wrapper, but it would be much simpler.

wdyt?